### DISCUSSION TITLE
sqlalchemy.orm.exc.DetachedInstanceError in unit test with socketio.test_client

### DISCUSSION THREAD

**Discussion by LostInDarkMath** [miguelgrinberg/Flask-SocketIO#1616] [Q&A]
**Describe the bug**
I just included `Flask-SocketIO` in my project which is based on `Flask-SQLAlchemy` and wrote some tests. But I in a simple test I get the following error:
```
sqlalchemy.orm.exc.DetachedInstanceError: Instance <User at 0x2812f970> is not bound to a Session; attribute refresh operation cannot proceed (Background on this error at: http://sqlalche.me/e/14/bhk3)
```
The error causing test can be reduced to the following:
```python
def test_receive_notifications_on_new_message_via_websockets(self) -> None:
    socketio_client = self.socketio.test_client(app=self.app, flask_test_client=self.client, headers=self.header())
    self._send_message(sender=self.dummy2, receiver=self.dummy)  # it fails here
```
If I comment out the the first line that creates the socketio test client, the errors does not appear. This is quite strange because I have no idea why the creation of socketio test client can lead to detached instances.
Do you have any idea whats going on here?

The dummy users `self.dummy` and `self.dummy2` are Wrappers for an ORM user class. The instances are created inside the `setup()` of the parent test class. I use them in almost any of my 300+ tests and I never had any problems before.

I hope you can help me here.
 
**Additional context**
I do not think that I use `Flask-SocketIO` wrong because if I start the server and my client everything works perfectly fine. Additionally, my other simple socketio tests also work perfectly:
```python
    def test_establish_websocket_connection_without_credentials(self) -> None:
        socketio_client = self.socketio.test_client(app=self.app, flask_test_client=self.client)
        self.assertFalse(socketio_client.is_connected())

    def test_establish_websocket_connection_successful_and_receive_initial_data(self) -> None:
        socketio_client = self.socketio.test_client(app=self.app, flask_test_client=self.client, headers=self.header())
        self.assertTrue(socketio_client.is_connected())
        res_list = socketio_client.get_received()
        self.assertEqual(1, len(res_list))
        # more assertions here
```

**Comment by miguelgrinberg** [comment]
Two important things are missing:

1. the complete stack trace of the error
2. the code that fails, which I assume is the `_send_message()` method of your unit test class?

**Comment by LostInDarkMath** [comment]
I did some further investigation on this and I think I found out why the instances got detached. In the `connect()` method of `test_client.py` you push and pop an flask app context with these lines:

```python
  with self.app.app_context():
            self.socketio.server._handle_eio_message(self.eio_sid,  pkt.encode())
```
But this calls the `session.remove()` in flask-SQLAlchemy:

```python
  @app.teardown_appcontext
        def shutdown_session(response_or_exc):
            if app.config['SQLALCHEMY_COMMIT_ON_TEARDOWN']:
                warnings.warn(
                    "'COMMIT_ON_TEARDOWN' is deprecated and will be"
                    " removed in version 3.1. Call"
                    " 'db.session.commit()'` directly instead.",
                    DeprecationWarning,
                )

                if response_or_exc is None:
                    self.session.commit()

            self.session.remove()
            return response_or_exc
```

This leads to the `DetachedInstanceError` in my case, because I create a User via ORM, then I connect the SocketIO test client and after that I read some properties of the ORM user.

So my question is: How can I solve this problem?

**Reply by miguelgrinberg** [reply]
I still haven't seen the code that fails.

**Reply by LostInDarkMath** [reply]
Unfortunately I not able to update the initial post. I have created a minimal example that reproduces the bug. I hope this helps you.
```python
from flask import Flask, request
from flask_socketio import SocketIO
from flask_sqlalchemy import SQLAlchemy

db = SQLAlchemy()
socketio = SocketIO(async_mode='eventlet')


class User(db.Model):
    id = db.Column(db.Integer, primary_key=True)
    username = db.Column(db.String(80))


def create_app():
    app = Flask(__name__)
    db.init_app(app)
    socketio.init_app(app)

    @socketio.on('connect')
    def on_connect():
        user_id = request.headers['id']
        user = User.query.get(user_id)
        user.username = 'donald'
        db.session.commit()

    return app
```

```python
from unittest import TestCase

from main import User, create_app, db, socketio


class MyTests(TestCase):
    def setUp(self) -> None:
        self.app = create_app()
        self.app_context = self.app.app_context()
        self.app_context.push()
        self.flask_client = self.app.test_client()
        db.create_all()

    def tearDown(self) -> None:
        db.session.remove()
        db.drop_all()
        db.get_engine(self.app).dispose()
        self.app_context.pop()

    def test_example(self):
        user = User(id=42, username='test')
        db.session.add(user)
        db.session.commit()

        socketio_client = socketio.test_client(
            app=self.app, flask_test_client=self.flask_client, headers={'id': user.id})
        self.assertTrue(socketio_client.is_connected())
        print(user.username)
```

This crashes with the following stacktrace:

```
FAILED (errors=1)

Error
Traceback (most recent call last):
  File "C:\Users\Willi\PycharmProjects\flask_socketio_bug\tests.py", line 28, in test_example
    print(user.username)
  File "C:\Users\Willi\AppData\Local\Programs\Python\Python38\lib\site-packages\sqlalchemy\orm\attributes.py", line 481, in _get_
    return self.impl.get(state, dict_)
  File "C:\Users\Willi\AppData\Local\Programs\Python\Python38\lib\site-packages\sqlalchemy\orm\attributes.py", line 926, in get
    value = self._fire_loader_callables(state, key, passive)
  File "C:\Users\Willi\AppData\Local\Programs\Python\Python38\lib\site-packages\sqlalchemy\orm\attributes.py", line 957, in _fire_loader_callables
    return state._load_expired(state, passive)
  File "C:\Users\Willi\AppData\Local\Programs\Python\Python38\lib\site-packages\sqlalchemy\orm\state.py", line 710, in _load_expired
    self.manager.expired_attribute_loader(self, toload, passive)
  File "C:\Users\Willi\AppData\Local\Programs\Python\Python38\lib\site-packages\sqlalchemy\orm\loading.py", line 1363, in load_scalar_attributes
    raise orm_exc.DetachedInstanceError(
sqlalchemy.orm.exc.DetachedInstanceError: Instance <User at 0x5037490> is not bound to a Session; attribute refresh operation cannot proceed (Background on this error at: https://sqlalche.me/e/14/bhk3)
```

The reason for the crash is, that `socketio.test_client()` pushes and pops a Flask app context. But when it is popped, it implicitly calls `session.remove()` which causes the `DetachedInstanceError`.

Do you need more information Miguel? Should I open a new issue for this? I think this is a bug in the socketIO test client because I do not think that this is desired behavior, right?.

**Reply by miguelgrinberg** [reply]
I'm not sure how to address this problem. Basically what you want is to have a database session that is shared between your test code and your Socket.IO endpoints. Fixing the problem would mean that any changes you make in the test code might affect endpoints that you invoke later, and likewise, changes made in the endpoints will automatically appear in the database operations you do in your test. So for example, if you modify a user in your test code and forget to commit it, the next `db.session.commit()` issued in a Socket.IO endpoint will also write that change to the database.

This is a really bad way to run tests, in my opinion. I don't think you should expect you have access to a long lived database session in your test.

**Reply by LostInDarkMath** [reply]
This is a bad way? But that's the way you did it in your Book **Flask Web Development**: https://github.com/miguelgrinberg/flasky/blob/master/tests/test_api.py#L110

I'm a bit confused now. I just did it the way you did it in your book.

**Reply by miguelgrinberg** [reply]
Well, do you understand what I'm saying? Do we agree that this is a bad idea? You haven't shown me actual code, so I can only comment on the example code you are showing me. The `print(user.username)` that you added in your test triggers a database query. Once the error is removed, this query is going to run under the context of a session that is not clean, the state of that session is going to be whatever the `connect` event handler in your application left it.

Flask is permissive in this respect, it does let you reuse sessions, and for that reason the test that you linked from my book's project works fine. Remember that I asked to see your code? This is why, depending on what you do, your usage might be okay (like my test) or not okay, and I wanted to see which is it.

As I said above, "I'm not sure how to address this problem". I can remove the error and let you share your database sessions left and right. That is going to make it easier to make mistakes though. It would be better to not reuse your database session.

**Reply by LostInDarkMath** [reply]
> This is why, depending on what you do, your usage might be okay (like my test).

I'm aware of this. The thing is that the shared session never caused any problems before. Like your `flasky` tests. Sure, uncommitted changes or forgotten rollbacks could cause problems. But in fact, these were never problems for me. So in my point of view the call to `session.remove()` is the problem here. The socketIO test client closes the session during connecting. I do not know if this is really necessary. But if not, I think that this should be removed. Currently it is just not possible to reuse the the `session` because the test client prevent that and do not think that it is a good thing to force your users to maintain multiple sessions.

**Reply by miguelgrinberg** [reply]
> I'm aware of this. The thing is that the shared session never caused any problems before.

You can say the same thing of all other bad practices. They never cause any problems until they do. :)

I'll think about how to address this problem. I hope you understand that I cannot base a solution on the fact that this has never been a problem "for you". I want to find a good solution, not something that works in your case but potentially creates issues for other users.

**Reply by LostInDarkMath** [reply]
Sure. I believe that there is a solution that does not cause any problems for others. I'm looking forward for your solution. Thanks for your time and your fast responses!

**Comment by LostInDarkMath** [comment]
One last question concerning this:
> It would be better to not reuse your database session.

Do you have an example how to do this? You said this is best practice but I dont find a way how to do this. There is also an open question on StackOverflow for this: https://stackoverflow.com/questions/44744183/how-to-isolate-sqlalchemy-session-between-tests-and-tested-code

**Reply by miguelgrinberg** [reply]
All you need to do is not use model instances that you obtained in a previous logical session.

For example, this is what you are doing in the test that you posted here as an example:

- Get a user (this starts a session and leaves it open)
- Invoke a Socket.IO endpoint (the session is already dirty, Flask-SocketIO before the fix closes it at the end)
- Use the user that you obtained in item 1, which is now detached because the session was closed.

A better way to do this would be:

- Get a user, do whatever you need to do with it and close the session
- Run the Socket.IO endpoint, now with a clean session
- Get the user again, and do what you need to do with it, close the session when you are done.

Note that SQLAlchemy 1.4 (which is not fully supported by Flask-SQLAlchemy) introduced context managers for sessions. This is going to help prevent this type of mistakes, because you have more control over the life of a session without having to manually close it.
